reference/ast-examples.md

```diff
@@ -0,0 +1,185 @@
+# AST Examples
+
+When building recipes, it's important to understand how OpenRewrite Abstract Syntax Trees (AST) correspond to code. You couldn't, for example, properly rename a variable with a recipe unless you knew that [J.Identifier](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L2310-L2343) is the class used to represent a variable.
+
+To help you get started on working with ASTs and OpenRewrite, this guide will:
+
+* [Explain how ASTs work at a high-level](#high-level-ast-explanation)
+* [Provide a sample chunk of code to demonstrate ASTs](#code)
+* Discuss how that code relates to different types of ASTs
+  * [Binary](#binary)
+  * [Block](#block)
+  * [ClassDeclaration](#classdeclaration)
+  * [CompilationUnit](#compilationunit)
+  * [Expression](#expression)
+  * [FieldAccess](#fieldaccess)
+  * [Identifier](#identifier)
+  * [MethodDeclaration](#methoddeclaration)
+  * [MethodInvocation](#methodinvocation)
+  * [NewClass](#newclass)
+  * [Statement](#statement)
+  * [VariableDeclarations](#variabledeclarations)
+* [Teach you how to use the debugger to learn more about ASTs yourself](#using-the-debugger-to-detect-asts)
+
+## High-level AST Explanation
+
+In order for a computer to parse and change code, there needs to be some data structure defined that can accurately represent said code. OpenRewrite uses an [Abstract Syntax Tree](/v1beta/abstract-syntax-trees.md) (AST) for this purpose. Like many tree data structures, ASTs are often composed of many other ASTs. 
+
+For instance, you could have a [CompilationUnit](#compilationunit) which then has a [ClassDeclaration](#classdeclaration) inside of that which then has a [MethodDeclaration](#methoddeclaration) inside of that and so on. On top of being a valid AST that would _also_ represent valid Java code. 
+
+On the other hand, you _could_ define a [Binary](#binary) with other Binaries inside of that. That would be a valid AST but it **would not** represent valid Java code as just entering `1+2` (a valid Binary) in a Java file would not compile (that code would need to be in an expression which is then in a class to be valid).
+
+> Every object in the [code section](#code) is an AST in the sense that it implements a Tree interface. These objects, however, need to be combined in an appropriate way to represent valid Java. It is your duty as a recipe author to ensure that the final AST represents valid Java code.
```

Comment:
```suggestion
For instance, a [ClassDeclaration](#classdeclaration) is an AST which defines a class. A typical class declaration will be composed of elements such as fields, methods, constructors, and inner classes. Each of these are represented as their own, smaller AST. So the term "AST" may refer to an entire, complete Java file or just one piece of it.

It is possible to manipulate ASTs to create code which will not compile. OpenRewrite provides some safeguards against grammatically invalid transformations in its type system - you can't replace an import statement with a method declaration, for example. But there are plenty ways ASTs can be valid according to the Java grammar without being valid, compilable programs. For example, nothing prevents you from modifying an AST such that a variable is used before it is defined. It is the responsibility of recipe authors to consider language semantics and the full range of possibilities when making changes. In accordance with the principle of Doing No Harm, always err on the side of leaving code untouched rather than making a risky change. 
```

Readers don't yet know what a `CompilationUnit` or `ClassDeclaration` are, so these terms cannot yet be used to explain related concepts.
A java programmer is likely to infer the meaning of `ClassDeclaration` correctly, but I had no idea that the name for an entire Java file was a "compilation unit" until I worked here.